sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,417 @@
+        self.db_info.clear()
+
+    def store_db_infos(self, db_infos):
+        for db_info in db_infos:
+            self.db_info[db_info['name']] = db_info
+
+    def store(self, db_name, schema, tables, columns_count):
+        self._columns_count += columns_count
+        schemas = self.db_to_schemas.setdefault(db_name, {})
+        if schema["id"] in schemas:
+            known_tables = schemas[schema["id"]].setdefault("tables", [])
+            known_tables = known_tables.extend(tables)
+        else:
+            schemas[schema["id"]] = schema
+            schemas[schema["id"]]["tables"] = tables
+
+    def columns_since_last_submit(self):
+        return self._columns_count
+
+    def truncate(self, json_event):
+        max_length = 1000
+        if len(json_event) > max_length:
+            return json_event[:max_length] + " ... (truncated)"
+        else:
+            return json_event
+
+    # NOTE: DB with no schemas is never submitted
+    def submit(self):
+        if not self.db_to_schemas:
+            return
+        self._columns_count = 0
+        event = {**self._base_event, "metadata": [], "timestamp": time.time() * 1000}
+        for db, schemas_by_id in self.db_to_schemas.items():
+            db_info = {}
+            if db not in self.db_info:
+                self._log.error("Couldn't find database info for %s", db)
+                db_info["name"] = db
+            else:
+                db_info = self.db_info[db]
+            event["metadata"] = event["metadata"] + [{**(db_info), "schemas": list(schemas_by_id.values())}]
+        json_event = json.dumps(event, default=default_json_event_encoding)
+        self._log.debug("Reporting the following payload for schema collection: {}".format(self.truncate(json_event)))
+        self._submit_to_agent_queue(json_event)
+        self.db_to_schemas.clear()
+
+
+def agent_check_getter(self):
+    return self._check
+
+
+class Schemas(DBMAsyncJob):
+
+    # Requests for infromation about tables are done for a certain amount of tables at the time
+    # This number of tables doesnt slow down performance by much (15% compared to 500 tables)
+    # but allows the queue to be stable.
+    TABLES_CHUNK_SIZE = 500
+    # Note: in async mode execution time also cannot exceed 2 checks.
+    MAX_EXECUTION_TIME = 10
+    MAX_COLUMNS_PER_EVENT = 100_000
+
+    def __init__(self, check, config):
+        self._check = check
+        self._log = check.log
+        self.schemas_per_db = {}
+        self._last_schemas_collect_time = None
+        collection_interval = config.schema_config.get('collection_interval', DEFAULT_SCHEMAS_COLLECTION_INTERVAL)
+        self._max_execution_time = min(
+            config.schema_config.get('max_execution_time', self.MAX_EXECUTION_TIME), collection_interval
+        )
+        e = is_affirmative(config.schema_config.get('enabled', True))
+        print(e)
+        super(Schemas, self).__init__(
+            check,
+            run_sync=True,
```

Comment:
should this be configurable? It might make sense for customers with larger schemas to want to run this in an async thread

Reply:
pls take a look at this answer https://github.com/DataDog/integrations-core/pull/17258#discussion_r1630173355